frontend/src/metabase/metadata/components/FieldOrderSidesheet/FieldOrderSidesheet.tsx

```diff
@@ -0,0 +1,117 @@
+import { PointerSensor, useSensor } from "@dnd-kit/core";
+import { useMemo, useState } from "react";
+import { useDeepCompareEffect } from "react-use";
+import { t } from "ttag";
+import _ from "underscore";
+
+import { Sidesheet } from "metabase/common/components/Sidesheet";
+import {
+  type DragEndEvent,
+  SortableList,
+} from "metabase/core/components/Sortable";
+import Tables from "metabase/entities/tables";
+import { useDispatch } from "metabase/lib/redux";
+import { PLUGIN_FEATURE_LEVEL_PERMISSIONS } from "metabase/plugins";
+import { Flex } from "metabase/ui";
+import type Table from "metabase-lib/v1/metadata/Table";
+import type { TableFieldOrder, TableId } from "metabase-types/api";
+import type { State } from "metabase-types/store";
+
+import { FieldOrderPicker } from "./FieldOrderPicker";
+import { SortableField } from "./SortableField";
+import { getId, getItems, getItemsOrder, sortItems } from "./lib";
+
+/**
+ * This is to prevent FieldOrderPicker's focus state outline being cut off.
+ * Mantine Button's outline-width is 2px.
+ */
+const BUTTON_OUTLINE_WIDTH = 2;
+
+interface OwnProps {
+  tableId: TableId;
+  isOpen: boolean;
+  onClose: () => void;
+}
+
+interface Props extends OwnProps {
+  table: Table;
+}
+
+const FieldOrderSidesheetBase = ({ isOpen, table, onClose }: Props) => {
+  const dispatch = useDispatch();
+  const pointerSensor = useSensor(PointerSensor, {
+    activationConstraint: { distance: 15 },
+  });
+  const initialItems = useMemo(() => getItems(table.fields), [table.fields]);
+  const initialOrder = useMemo(
+    () => getItemsOrder(initialItems),
+    [initialItems],
+  );
+  const [items, setItems] = useState(initialItems);
+  const [order, setOrder] = useState(initialOrder);
+  const sortedItems = useMemo(() => sortItems(items, order), [items, order]);
+  const isDragDisabled = sortedItems.length <= 1;
+
+  const handleSortEnd = ({ itemIds }: DragEndEvent) => {
+    setOrder(itemIds);
+    dispatch(Tables.actions.setFieldOrder(table, itemIds));
+  };
+
+  const handleFieldOrderChange = (value: TableFieldOrder) => {
+    dispatch(Tables.actions.updateProperty(table, "field_order", value));
+  };
+
+  useDeepCompareEffect(() => {
```

Comment:
If you watch the "before" demo closely, right after performing drag and drop the items will shortly flicker back to the original order, and then back to the new one.
These deep compare effects help avoid that.
Fields are mapped to items because using a field as a dep here results in an infinite loop.

I will add this comment in code before merging the PR.

Reply:
be9ac259e3e7a788b8c20fef9971325d64f7bf30